### DISCUSSION TITLE
大家开发iLogtail有什么好用的VSCode插件推荐？

### DISCUSSION THREAD

**Discussion by yyuuttaaoo** [alibaba/ilogtail#299] [Show and tell]
### 插件名称

Remote Development

### 插件链接

https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.vscode-remote-extensionpack

### 推荐理由

使用远程开发机的必备插件，支持通过ssh链接到远端。最近还发现了可以直接使用一个镜像生成container作为开发环境，对于一个团队统一开发环境，新人快速onboard简直神器。

### 使用方法

Shift + Command + P（Mac）或Ctrl + Shift + P（Win）打开命令面板，输入reopen，选择Remote-Containers: Reopen in Container。
![1659944991094-2683623d-3165-4953-8fa6-82f2c56917fe](https://user-images.githubusercontent.com/1827594/183560708-7c143707-3f0c-411c-a655-49d4bc55cd7e.png)

**Comment by messixukejia** [comment]
当管理多个项目的时候，VS code切换比较麻烦，每次都得通过路径打开。

推荐使用[Project Manager](https://marketplace.visualstudio.com/items?itemName=alefragnani.project-manager)插件，可以方便进行多项目管理，而且还支持tag分组。
![image](https://user-images.githubusercontent.com/13250519/185014581-e491551a-2138-4096-979d-3ce01d839353.png)

**Comment by yyuuttaaoo** [comment]
插件名称
Inline Parameters Extended for VSCode

插件链接
https://marketplace.visualstudio.com/items?itemName=RobertOstermann.inline-parameters-extended

推荐理由
自动为Go语言增加参数名提示，有助代码阅读，类似IntelliJ。

使用方法
![image](https://github.com/alibaba/ilogtail/assets/1827594/536db6d3-a128-4ccb-9106-513a34a3c93e)

**Reply by yyuuttaaoo** [reply]
似乎不需要这个插件了，现在C++和Go Extension都支持inlay hint了
![image](https://github.com/alibaba/ilogtail/assets/1827594/3f73d421-a108-424d-a034-d3245d9a3de2)

**Comment by cyenxchen** [comment]
Name: CodeLLDB
Id: vadimcn.vscode-lldb
Description: A native debugger powered by LLDB.  Debug C++, Rust and other compiled languages.
Version: 1.9.2
Publisher: Vadim Chugunov
VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=vadimcn.vscode-lldb

推荐理由
使用该插件可以利用vscode的UI调试界面调试ilogtail的核心代码，配合对应的编译脚本就可以一键调试ilogtail核心代码。

**Reply by kabitionno** [reply]
你好，如何操作的，能给一个说明文档，谢谢

**Reply by cyenxchen** [reply]
ilogtail的底层基础镜像还没升级呢，弄起来很麻烦，等官方升级之后我会提交个文档的。

**Reply by kabitionno** [reply]
# 整理了一下步骤可以界面单步调试了
1. windows 下安装 wsl2 环境的 ubuntu20.04
2. 停用 vscode c/c++ 和 ### C++ Extension 插件
3. 安装 clangd 和 CodeLLDB 插件，根据vscode提示安装clangd server
4. vscode终端安装 GLBC-2.18(CodeLLDB需要2.18+)
查看glibc版本 strings /lib64/libc.so.6 |grep GLIBC_
```bash
wget http://mirrors.ustc.edu.cn/gnu/libc/glibc-2.18.tar.gz
tar -zxvf glibc-2.18.tar.gz
cd glibc-2.18
mkdir build
cd build
../configure --prefix=/usr
make -sj$(nproc)
sudo make install
```
5. vscode命令行编译带debug信息的可执行代码
```bash
cd /workspaces/ilogtail
mkdir -p core/build
cd core/build
cmake -D CMAKE_BUILD_TYPE=Debug ..
make -sj$(nproc)
```
6.  vscode命令行使用CMake生成compile_command.json
CMake在指定构建目录的命令中加入以下选项就可以生成compile_command.json (如果你不熟悉CMake, 后文会介绍个人使用的最简CMake项目模板与使用指南)
```bash
cd /workspaces/ilogtail/core
cmake -B build  -DCMAKE_EXPORT_COMPILE_COMMANDS=1
```
7. [文件]-[首选项]-[设置] 搜索 clangd 添加 --compile-commands-dir=${workspaceFolder}/core/build
8. 项目根目录下的.vscode/launch.json
ilogtail主要由两部分(c++\go)组成且可以通过如下配置分开调试
```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "lldb",
            "request": "launch",
            "name": "Debug c++",
            "program": "${workspaceFolder}/core/build/ilogtail",
            "args": [],
            "cwd": "${workspaceFolder}"
        },
        {
            "name": "Launch Package",
            "type": "go",
            "request": "launch",
            "mode": "debug",
            "program": "${workspaceFolder}/plugin_main",
            "env": {
                "LD_LIBRARY_PATH": "${workspaceFolder}/pkg/logtail"
            }
        }
    ]
}
```
9. vscode 运行与调试 切换到对应启动项后, 按 F5 开始debug,此时ui上会出现如下控制悬浮框

**Reply by cyenxchen** [reply]
我之前弄的时候只升级glibc-2.18版本是没用的，我记得还得手动改好些东西，比较麻烦，你这边直接升级glibc的库就可以用了吗？
另外，创建目录，编译这些命令可以放到task里面，那样就可以真正做到一键编译调试了，task.json内容如下（纯手撸还没测试，但大概思路没问题）：
```json
{
    // See https://go.microsoft.com/fwlink/?LinkId=733558
    // for the documentation about the tasks.json format
    "version": "2.0.0",
    "tasks": [
        {
            // 创建目录
            "label": "mkdir",
            "type": "shell",
            "command": "mkdir",
            "args": ["-p","${workspaceFolder}/core/build"]
        },
        {
            // 执行CMake
            "label": "cmake",
            "type": "shell",
            "command": "cmake",
            "args": ["-D","CMAKE_BUILD_TYPE=Debug", "CMAKE_EXPORT_COMPILE_COMMANDS=1", "../"],
            "dependsOn":["mkdir"],
            "options": {
                "cwd": "${workspaceFolder}/core/build"
            }
        },
        {
            // 编译
            "label": "build",
            "type": "shell",
            "command": "make",
            "args": ["j","$(nproc)"],
            "dependsOn":["cmake"],
            "options": {
                "cwd": "${workspaceFolder}/core/build"
            }
        },
        {
            // 完成动态库的拷贝
            "label": "cp",
            "type": "shell",
            "command": "cp",
            "args": ["-a", "./ilogta", "./plugin/libPluginAdapter.so", "${workspaceFolder}/output"],
            "dependsOn":["build"],
            "options": {
                "cwd": "${workspaceFolder}/core/build"
            }
        }
    ]
}
```
接着就可以在launch.json中使用这个task了:
```json
{
    ......
    "preLaunchTask": "cp",
    ......
}
```

**Reply by kabitionno** [reply]
我用的是官方1.7版本的镜像，目前是可以的，如果不行可以考虑  sudo yum install clangd lldb ，反正这步我之前试过也不知道其作用没

**Reply by messixukejia** [reply]
👍🏻。可以写篇文章分享下。https://github.com/alibaba/ilogtail/tree/main/docs/cn/awesome-ilogtail

**Comment by byte-sourcerer** [comment]
devcontainer 镜像是不是应该考虑加个 clangd?🤔

**Reply by yyuuttaaoo** [reply]
> devcontainer 镜像是不是应该考虑加个 clangd?🤔

加了

**Comment by beidou2018** [comment]
有没有研究过使用jetbrain clion去远程启动服务器开发容器的？？？
